provider/resource_keycloak_identity_provider_token_exchange_scope_permission.go

```diff
@@ -0,0 +1,292 @@
+				Computed:    true,
+				Description: "Resource id representing the identity provider, this automatically created by keycloak",
+			},
+			"authorization_token_exchange_scope_permission_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Permission id representing the Permission with scope 'Token Exchange' and the resource 'authorization_idp_resource_id', this automatically created by keycloak, the policy id will be set on this permission",
+			},
+		},
+	}
+}
+
+func setIdentityProviderTokenExchangeScopePermissionClientPolicy(keycloakClient *keycloak.KeycloakClient, realmId, providerAlias string, clients []string) error {
+	identityProviderPermissions, err := keycloakClient.GetIdentityProviderPermissions(realmId, providerAlias)
+	if err != nil {
+		return err
+	}
+
+	realmManagementClient, err := keycloakClient.GetOpenidClientByClientId(realmId, "realm-management")
+	if err != nil {
+		return err
+	}
+
+	tokenExchangeScopedPermissionId, err := identityProviderPermissions.GetTokenExchangeScopedPermissionId()
+	if err != nil {
+		return err
+	}
+
+	permission, err := keycloakClient.GetOpenidClientAuthorizationPermission(realmId, realmManagementClient.Id, tokenExchangeScopedPermissionId)
+	if err != nil {
+		return err
+	}
+
+	if len(permission.Policies) == 0 {
+		policyId, err := createClientPolicy(keycloakClient, realmId, realmManagementClient.Id, providerAlias, clients)
+		if err != nil {
+			return err
+		}
+		permission.Policies = []string{policyId}
+		return keycloakClient.UpdateOpenidClientAuthorizationPermission(permission)
+
+	} else if len(permission.Policies) == 1 {
+		openidClientAuthorizationClientPolicy, err := keycloakClient.GetOpenidClientAuthorizationClientPolicy(realmId, realmManagementClient.Id, permission.Policies[0])
+		if err != nil {
+			return err
+		}
+		openidClientAuthorizationClientPolicy.Clients = clients
+		return keycloakClient.UpdateOpenidClientAuthorizationClientPolicy(openidClientAuthorizationClientPolicy)
+
+	} else {
+		return fmt.Errorf("only supports 1 client policy is supporterd on the identity provider token exchange scope based permission, but %d were found", len(permission.Policies))
+	}
+}
+
+func createClientPolicy(keycloakClient *keycloak.KeycloakClient, realmId, realmManagementClientId, providerAlias string, clients []string) (string, error) {
+	openidClientAuthorizationClientPolicy := &keycloak.OpenidClientAuthorizationClientPolicy{
+		RealmId:          realmId,
+		ResourceServerId: realmManagementClientId,
+		Name:             realmId + "_" + providerAlias + "_client_policy",
+		DecisionStrategy: "UNANIMOUS",
+		Logic:            "POSITIVE",
+		Type:             "client",
+		Clients:          clients,
+	}
+	err := keycloakClient.NewOpenidClientAuthorizationClientPolicy(openidClientAuthorizationClientPolicy)
+	if err != nil {
+		if strings.Contains(err.Error(), "409 Conflict") {
+			index := 1
+			for index <= 10 && err != nil {
```

Comment:
can you help me understand why we're trying to do this 10 times?  it looks like we're avoiding an error due to the policy name already being in use - can we let users decide what name they want to use instead?

Reply:
well, I did not know how to solve this. But I a user manually disables the Idp permissions,
The policy stays behind. so you would never be able to recreate the policy. So even If you let the user choose the policy name. the CreateAfterManualDestroyTest would fail.
Furthermore If an user can choose their own policyName there might be confusion that they have to create a policy outside of this resource and link to it.....
I know, it still not bullet proof

Reply:
I am gonna remove the index and 10 times loop and just add random string in case the generated name random already existed
What do you think?

Reply:
Yeah this is probably fine.  My understanding is that this can only happen when users manually make changes outside of Terraform, so I am not too worried about it.

Reply:
Yes exactly